avm/src/daemon/ctrl.rs

```diff
@@ -611,10 +655,21 @@ impl ControlPort {
   }
 
   pub async fn update_vms(self: &mut ControlPort, self_ip: &str, vms: Vec<VMMetadata>) {
-    let ips = vms
+    let ips: Vec<String> = vms
       .iter()
       .map(|vm| vm.private_ip_addr.to_string())
       .collect();
+    // Detect changes and exit early if nothing has changed
+    let changed = if ips.len() != self.vms.len() {
+      true
+    } else if ips.iter().any(|ip| !self.vms.contains_key(ip)) {
+      true
+    } else {
+      false
+    };
```

Comment:
nit: all of this can be
```suggestion
    let changed = ips.len() != self.vms.len() || ips.iter().any(|ip| !self.vms.contains_key(ip));
```

Reply:
:facepalm: